keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,465 @@
+nitty-gritty.
+-->
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+The `NodeAddress` of type `NodeHostName` will be moved to a new
+`Node.Status.Hostname` field.
+
+```
+<<[UNRESOLVED hostname ]>>
+Is `Node.Status.Hostname` required or optional? Check existing constraints.
+<<[/UNRESOLVED]>>
+```
+
+The other information in `Node.Status.Addresses` will be moved to a
+new `Node.Status.IPs` array of type `NodeIP`:
+
+    // NodeIP provides information about a single Node IP address
+    type NodeIP struct {
+            // IP is a node IP address
+            IP string `json:"ip"`
+
+            // Scope gives the scope of IP
+            // +optional
+            Scope NodeIPScope `json:"scope,omitempty"`
+
+            // DNSName is a DNS name that resolves to IP
+            // +optional
+            DNSName string `json:"dnsName,omitempty"`
+    }
+
+    // NodeIPScope describes the scope within which a Node IP exists
+    type NodeIPScope string
+
+    const (
+            // NodeIPUnknown indicates an IP address with unknown scope. (It is at least
+            // reachable by all other nodes within the cluster, but may or may not be
+            // accessible beyond that.)
+            NodeIPUnknown NodeIPScope = ""
+
+            // NodeIPInternal indicates an IP address that exists only within the scope of
+            // some private network (eg, an intranet or private cloud network).
+            NodeIPInternal NodeIPScope = "Internal"
+
+            // NodeIPExternal indicates an IP address that exists only outside the
+            // cluster. (eg, traffic sent to this address from inside the cluster may
+            // leave the cluster and then come back with the destination IP rewritten to a
+            // Internal IP address.)
+            NodeIPExternal NodeIPScope = "External"
+
+            // NodeIPPublic indicates an IP address that is expected to be directly
+            // reachable from anywhere, inside or outside the cluster.
+            NodeIPPublic NodeIPScope = "Public"
```

Comment:
I have doubts about this `NodeIPPublic`option, it has a lot of connotations :/, 

> an IP address that is expected to be directly reachable from anywhere

Reply:
The existing options have a lot of connotations too. eg, this is partly inspired by https://github.com/kubernetes/kubernetes/pull/86918#discussion_r363997750.

Note that it's "*expected* to be directly reachable from anywhere", not "required to be directly reachable from anywhere". And if the cloud provider doesn't know, it should just use `Unknown`

Reply:
> but IPv6 node IPs will most likely be `Public`, and bare-metal IPs will generally be `Unknown`

I think that continuing with the IPv6 analogy, we can merge `Unknown` and `External` in `Global`, taking a more minimalistic approach and avoiding the divergence on the use of IPs scopes between private/public cloud providers as you describe below. However, I don't know if the goal is exactly making that difference explicit ....

Reply:
The conversation in 86918 was specifically about the distinction between node IPs where traffic is expected to go directly to the node vs node IPs where traffic is expected to go to some piece of cloud infrastructure first, and a concern that we might accidentally end up preferring an IPv6 address of the latter type over an IPv4 address of the former type. The `Public` vs `External` distinction is exactly what was needed there.

I had also previously thought about just having some boolean flags on `NodeIP` instead of a type, like:

    Public   bool // IP is assumed to be globally reachable
    Indirect bool // Packets sent to this IP will be rewritten to another IP

So `Internal` would be `!public && !indirect`, `External` would be `public && indirect` and `Public` would be `public && !indirect`. But then that leaves `!public && indirect` unused, and it doesn't allow expressing unknown public-ness, which is really usually the case for autodetected bare metal IPs.

Reply:
| NodeIP Scope  |  Private | Public  |
|---|---|---|
| **Belongs Cluster** |Internal  | Public   |
| **Does not belong**  | Unknown  | External  | 

:+1: I got it :smile:

Reply:
> I got it

not quite; "Private" + "Does not belong" should be empty because there's no name that means that explicitly. `Unknown` means unknown. Eg, on bare metal, kubelet knows the local IPs aren't `External` but if they're not a known private IP range, then it has no way of knowing if they're `Internal` (ie, firewalled off from the internet) or `Public`. So it would call them `Unknown`.

Reply:
I think we're designing in a vacuum, hence why it's hard to pin down what semantics are useful.  What are the use cases for anything other than "reachable by all other nodes within the cluster"?

(Assuming we can even find more than one use case) Perhaps we should have different fields for the different use cases rather than capturing all reachability/cost/policy semantics in a single intrinsic "type"? (ie: "use this address when _doing_ foo-type operation", rather than "this address _is intrinsically_ foo-type")